### ISSUE TITLE
Add FIR filter examples to scipy.signal documentation

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#4368]
Here are a few examples that could easily be used to improve the docstrings and/or tutorial of scipy.signal: 

http://www.scipy.org/Cookbook/FIRFilter
http://www.scipy.org/Cookbook/FiltFilt
http://mpastell.com/2010/01/18/fir-with-scipy/ (IIRC @mpastell was OK with using that content, but I can't find an archived email that says that anymore. Matti can you confirm that this is OK?)

`signal.filtfilt` now has at least one nice example thanks to @WarrenWeckesser, and the tutorial is slowly improving mainly due to @endolith's PRs, but there's a ways to go....

**Comment by WarrenWeckesser** [comment]
@mpastell's examples are nice, but it is not true (perhaps better to say *no longer* true) that "SciPy does not have a function for directly designing a highpass FIR filter".  Check out the docstring for `scipy.signal.firwin`.   Admittedly, the examples in that docstring could use a little work; e.g. `f` is used but it isn't defined.

**Comment by WarrenWeckesser** [comment]
`<pedantic>`By the way, one wouldn't normally use `filtfilt` with a FIR filter.  A FIR filter is usually designed to have linear phase, and if the filter has linear phase, there is no advantage to applying the filter with `filtfilt`.`</pendantic>`

**Comment by rgommers** [comment]
</pendantic>

typo :)

**Comment by rgommers** [comment]
but yes, thanks for that remark.

**Comment by WarrenWeckesser** [comment]
> typo :)

Stop being so pendantic.

**Comment by mpastell** [comment]
Yes, feel free to use the examples and modify them in anyway you want. I'm happy if they are useful to someone.
There is slightly more up to date version in Pweave examples: http://mpastell.com/pweave/examples/FIR_design.html (regarding FIR highpass), I should probably update the old blog post too.

**Comment by WarrenWeckesser** [comment]
What I should have said first:  Yes, this is a great idea!

I put some `remez` examples here: https://gist.github.com/WarrenWeckesser/67bef61f496080aeb798

I have several other tasks to finish on my scipy to-do list before I'd work on this, so if anyone wants to expand the `remez` docstring with examples like those, feel free to use whatever you like.

**Comment by endolith** [comment]
Here's another example of filtfilt vs lfilter, though very similar to the one you linked to: http://dsp.stackexchange.com/a/19086/29

Could `firwin` be modified to take a btype argument so you don't need to do it by hand, similar to the IIR filter design functions?  Or make another function for that purpose that wraps it?

**Comment by WarrenWeckesser** [comment]
FWIW, here's another example that I created for a stackoverflow answer: 
http://stackoverflow.com/questions/16301569/bandpass-filter-in-python/16306891#16306891

**Comment by sdpan21** [comment]
I want to work on this issue but i am new in open source so I have no idea how to start working.  how should I get the source code in which the issue has been raised. please help.

**Comment by larsoner** [comment]
Maybe this tutorial I wrote for MNE could be adapted:

https://martinos.org/mne/stable/auto_tutorials/plot_background_filtering.html

At least some (non-signal-processing engineer) neuroscientists I've talked to have found it to be a useful intro. Of course for SciPy it could be expanded to deal with other design methods (e.g., `remez`).

**Comment by rgommers** [comment]
gh-8744 just improved the `remez` docstring, there are more functions discussed here - reopening

**Comment by Adarsh-NP** [comment]
Hi all, I think firwin2 should have more examples, especially for a custom filter. I want to add this something like this, but wanted to confirm from the maintainers and core members if the examples need to be from some standard reference material. 

I would like to know if it is correct or not, I tried replicating one of the examples that I encountered in my course but could not find in scipy.


![Screenshot from 2022-02-06 14-38-40](https://user-images.githubusercontent.com/54997561/152674367-5a68e645-0904-4ac3-8dc8-46ea46c2c234.png)

Additionally, I am confused if there is any difference between firwin and firwin2. I could notice that firwin2 accomodates type III and IV filters and I think we can give an arbitrary gain frequency map to it which is not possible in firwin. Kindly clarify if I got this right.

**Comment by Codeknightjl** [comment]
@Adarsh-NP I know this response is quite late but can you please provide more details on the example that you tried to replicate?

As for your final question,

> Additionally, I am confused if there is any difference between firwin and firwin2

The main difference is in the desired properties of the filter.

- firwin allows you to design a FIR filter by giving it the desired cutoff frequency.
- firwin2 allows you to design it by giving it a list of frequency points where the gains must be the corresponding values given by you in a separate list.

You are also correct that type III/IV filters can be designed using firwin2 but not with firwin. This line from the firwin documentary makes it clear.

> This function computes the coefficients of a finite impulse response filter. The filter will have linear phase; it will be Type I if numtaps is odd and Type II if numtaps is even.

Please let me know if I have missed anything.

**Comment by bdas123** [comment]
Is this issue still active?

**Comment by DietBru** [comment]
PR #23621 significantly improves the example section of the `firwin` docstr. Any review comments are, as always, very welcome.

**Comment by LeonxLJX** [comment]
I'd like to take this one (`DOC: signal: add FIR filter examples`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)
